File: 1 лаба/main.py
from datetime import datetime
import math
import matplotlib.pyplot as plt

step = 0.1                         # шаг для нахождения интервалов

x_array = []                       # массивы с координатами
y_array = []

intervals = []                     # интервалы, где есть корни
start_interval = end_interval = 0

a = b = func_a = func_b = 0        # точки интервалов
x = func_x = 0                     # среднее значение a и b
eps = 0.01                        # епсилент - минимальная погрешность
iterations = 0                     # количество шагов для нахождение результата

def my_function(x):
    return 4 * (1 + pow(x, 0.5)) * math.log(x) - 1

### Заполняем массивы x и y
counter = 0.01
while counter < 11:
    x_array.append(counter)
    y_array.append(my_function(counter))
    counter += 0.01

# Находим интервалы, где есть корни
start_time = datetime.now()

for item in range(1, len(y_array)):
    # от нижней точки к верхней и от верхней точки к нижней
    if y_array[item - 1] * y_array[item] < 0:
        start_interval = x_array[item - 1]
        end_interval = x_array[item]
        intervals.append([start_interval, end_interval])

# Находим приближенные корни
for interval in intervals:
    a = interval[0]
    b = interval[1]

    a_func = my_function(a)
    b_func = my_function(b)

    if a_func * b_func < 0:
        while (b - a) > abs(eps):
            iterations += 1
            x = (a + b) / 2
            func_x = my_function(x)
            
            if func_x == 0:
                print("f(x) - корень")
                break

            else:
                if func_a * func_x > 0:
                    a = x

                else:
                    b = x

        x = (a + b) / 2
        print("x:", x, "eps:", eps, "steps:", iterations)

    # Исключения с корнями
    elif a_func * b_func == 0:
        if func_a == 0:
            print("f(A) - корень")

            if func_b == 0:
                print("f(B) - корень")
                break
            break

        else:
            if func_b == 0:
                print("f(B) - корень")
                break

    else:
        print("Координаты Y интервалов либо больше 0, либо меньше")

print("time:", datetime.now() - start_time)

plt.title('Графики функций')
plt.xlabel("x")
plt.ylabel("y")
plt.grid()
plt.plot(x_array, y_array, label="1")
plt.legend()
plt.show()

File: 1 лаба/main.cpp
#include <iostream>
#include <cmath>
#include <ctime>

using namespace std;
//Функция для полинома:
double myFunction(double x)
{
    // return sin(x);
    return 4 * (1 + pow(x, 0.5)) * log(x) - 1;
}

//Функция поиска корня:
double findRoot(double (*f)(double), double a, double b, double eps)
{

    double x;
    int steps = 0;

    while ((b - a) / 2 > eps)
    {
        steps++;
        x = (a + b) / 2;
        if ((f(a) * f(x)) > 0)
            a = x;
        else
            b = x;
    }

    cout << "x: " << x << endl;
    cout << "f(x): " << f(x) << endl;
    cout << "eps: " << eps << endl;
    cout << "steps: " << steps << endl;
    steps = 0;
    return x;
}

double teoreticSteps(double a, double b, double eps) {
    double k = (log(b - a) - log(eps)) / log(2) + 1;
    return floor(k);
}

int main()
{
    //Интервал, погрешность и корень:
    double a, b, eps = 0.01;
    double x_array[100], y_array[100], intervals;
    double counter = 0.1;
    int start_time, end_time;
    // Заполняем массивы:
    for (int i = 0; i < 100; i++) {
        x_array[i] = counter;
        y_array[i] = myFunction(counter);
        // cout << "i: " << i <<
        // " x: " << x_array[i] <<
        // " y: " << y_array[i] << endl;
        counter += 0.1;
    }
    // Ищем корни в интервалах:
    for (int j = 1; j < sizeof(y_array) / sizeof(y_array[0]); j++) {
        if (y_array[j - 1] * y_array[j] < 0) {
            a = x_array[j - 1];
            b = x_array[j];

            start_time = clock();
            findRoot(myFunction, a, b, eps);
            end_time = clock() - start_time;
            
            cout << "time: " << (float)end_time / CLOCKS_PER_SEC << endl;
            cout << "teoretic steps: " << teoreticSteps(a, b, eps) << "\n\n" << endl;
        }
    }

    // system("python main.py");
}